lib/beacon/content.ex

```diff
@@ -1419,4 +1432,93 @@ defmodule Beacon.Content do
       _error -> :error
     end
   end
+
+  # PAGE VARIANTS
+
+  @doc """
+  Returns an `%Ecto.Changeset{}` for tracking variant changes.
+
+  ## Example
+
+      iex> change_page_variant(page_variant, %{name: "Variant A"})
+      %Ecto.Changeset{data: %PageVariant{}}
+
+  """
+  @doc type: :page_variants
+  @spec change_page_variant(PageVariant.t(), map()) :: Ecto.Changeset.t()
+  def change_page_variant(%PageVariant{} = page, attrs \\ %{}) do
+    PageVariant.changeset(page, attrs)
+  end
+
+  @doc """
+  Creates a new page variant and returns the page with updated variants association.
+  """
+  @doc type: :page_variants
+  @spec create_variant_for_page(Page.t(), %{name: binary(), template: binary(), weight: integer()}) ::
+          {:ok, Page.t()} | {:error, Changeset.t()}
+  def create_variant_for_page(page, attrs) do
+    changeset =
+      page
+      |> Ecto.build_assoc(:variants)
+      |> PageVariant.changeset(attrs)
+
+    Repo.transact(fn ->
+      with {:ok, %PageVariant{}} <- Repo.insert(changeset),
+           %Page{} = page <- Repo.preload(page, :variants, force: true),
+           %Page{} = page <- Lifecycle.Page.after_update_page(page) do
+        {:ok, page}
+      end
+    end)
+  end
+
+  @doc """
+  Updates a page variant and returns the page with updated variants association.
+  """
+  @doc type: :page_variants
+  @spec update_variant_for_page(Page.t(), PageVariant.t(), map()) :: {:ok, Page.t()} | {:error, Changeset.t()}
+  def update_variant_for_page(page, variant, attrs) do
+    changeset = PageVariant.changeset(variant, attrs)
+
+    Repo.transact(fn ->
+      with {:ok, _} <- validate_variant_template(changeset, page),
```

Comment:
This was intentional but I agree it could be handled better.  Perhaps `{:ok, ^changeset}` would make it clear that the changeset is being returned unchanged in the success case